### PR TITLE
Adding support for executing functions from referenced assemblies, to the source generated executor implementation.

### LINKED ISSUE #2030: Source gen function executor does not work when functions are defined in referenced assemblies
Source gen version of function executor does not work when functions are defined in referenced assemblies.

This issue happens when function app is using  the Azure.Functions.Worker.Sdk package version 1.16.0. 

#### Mitigation

To mitigate the issue, add the below entry to your `.csproj` file.

```xml
<PropertyGroup>
  <FunctionsEnableExecutorSourceGen>False</FunctionsEnableExecutorSourceGen>
</PropertyGroup>
```

### PR DESCRIPTION
**Pull request by kshyju**
Adding support for executing functions from referenced assemblies, to the source generated executor implementation.

Fixes #2030

- Excluding any non public methods when discovering valid azure functions (this change applies to the metadata provider as well).
- Public methods defined inside internal classes from referenced assemblies are using existing reflection based default executor.

#### Sample code
When tested on a function app which has:
- 1 http function (`Function1`) defined in the project.
- Has a nuget reference to `Microsoft.Azure.Functions.Worker.Extensions.OpenApi` 1.5.1 package, which will add 4 functions.
- Has a project reference which has 3 functions.
        -  1 Internal class with 2 functions (1 public method(`FooLibraryInternalClassPublicMethod1`) and 1 internal method(`FooLibraryInternalClassInternalMethod2`)
        -  1 Public class with 1 public method (`FooLibraryPublicClassPublicMethod1`).
```csharp
internal class DirectFunctionExecutor : IFunctionExecutor
{
    private readonly IFunctionActivator _functionActivator;
    private Lazy<IFunctionExecutor> _defaultExecutor;
    private readonly Dictionary<string, Type> types = new()
    {
        { "FunctionApp172.Function1", Type.GetType("FunctionApp172.Function1, FunctionApp172, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null")! },
        { "Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger", Type.GetType("Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger, Microsoft.Azure.Functions.Worker.Extensions.OpenApi, Version=1.0.0.0, Culture=neutral, PublicKeyToken=9475d07f10cb09df")! },
        { "ClassLibrary8.FooLibraryInternalClass", Type.GetType("ClassLibrary8.FooLibraryInternalClass, ClassLibrary8, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null")! },
        { "ClassLibrary8.FooLibraryPublicClass", Type.GetType("ClassLibrary8.FooLibraryPublicClass, ClassLibrary8, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null")! }
    };

    public DirectFunctionExecutor(IFunctionActivator functionActivator)
    {
        _functionActivator = functionActivator ?? throw new ArgumentNullException(nameof(functionActivator));
    }

    /// <inheritdoc/>
    public async ValueTask ExecuteAsync(FunctionContext context)
    {
        var inputBindingFeature = context.Features.Get<IFunctionInputBindingFeature>()!;
        var inputBindingResult = await inputBindingFeature.BindFunctionInputAsync(context)!;
        var inputArguments = inputBindingResult.Values;
        _defaultExecutor = new Lazy<IFunctionExecutor>(() => CreateDefaultExecutorInstance(context));

        if (string.Equals(context.FunctionDefinition.EntryPoint, "FunctionApp172.Function1.Run", StringComparison.Ordinal))
        {
            var instanceType = types["FunctionApp172.Function1"];
            var i = _functionActivator.CreateInstance(instanceType, context) as global::FunctionApp172.Function1;
            context.GetInvocationResult().Value = i.Run((global::Microsoft.Azure.Functions.Worker.Http.HttpRequestData)inputArguments[0]);
        }
        else if (string.Equals(context.FunctionDefinition.EntryPoint, "Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger.RenderSwaggerDocument", StringComparison.Ordinal))
        {
            var instanceType = types["Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger"];
            var i = _functionActivator.CreateInstance(instanceType, context) as global::Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger;
            context.GetInvocationResult().Value = await i.RenderSwaggerDocument((global::Microsoft.Azure.Functions.Worker.Http.HttpRequestData)inputArguments[0], (string)inputArguments[1], (global::Microsoft.Azure.Functions.Worker.FunctionContext)inputArguments[2]);
        }
        else if (string.Equals(context.FunctionDefinition.EntryPoint, "Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger.RenderOpenApiDocument", StringComparison.Ordinal))
        {
            var instanceType = types["Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger"];
            var i = _functionActivator.CreateInstance(instanceType, context) as global::Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger;
            context.GetInvocationResult().Value = await i.RenderOpenApiDocument((global::Microsoft.Azure.Functions.Worker.Http.HttpRequestData)inputArguments[0], (string)inputArguments[1], (string)inputArguments[2], (global::Microsoft.Azure.Functions.Worker.FunctionContext)inputArguments[3]);
        }
        else if (string.Equals(context.FunctionDefinition.EntryPoint, "Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger.RenderSwaggerUI", StringComparison.Ordinal))
        {
            var instanceType = types["Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger"];
            var i = _functionActivator.CreateInstance(instanceType, context) as global::Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger;
            context.GetInvocationResult().Value = await i.RenderSwaggerUI((global::Microsoft.Azure.Functions.Worker.Http.HttpRequestData)inputArguments[0], (global::Microsoft.Azure.Functions.Worker.FunctionContext)inputArguments[1]);
        }
        else if (string.Equals(context.FunctionDefinition.EntryPoint, "Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger.RenderOAuth2Redirect", StringComparison.Ordinal))
        {
            var instanceType = types["Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger"];
            var i = _functionActivator.CreateInstance(instanceType, context) as global::Microsoft.Azure.Functions.Worker.Extensions.OpenApi.DefaultOpenApiHttpTrigger;
            context.GetInvocationResult().Value = await i.RenderOAuth2Redirect((global::Microsoft.Azure.Functions.Worker.Http.HttpRequestData)inputArguments[0], (global::Microsoft.Azure.Functions.Worker.FunctionContext)inputArguments[1]);
        }
        else if (string.Equals(context.FunctionDefinition.EntryPoint, "ClassLibrary8.FooLibraryInternalClass.Run1", StringComparison.Ordinal))
        {
            await _defaultExecutor.Value.ExecuteAsync(context);
        }
        else if (string.Equals(context.FunctionDefinition.EntryPoint, "ClassLibrary8.FooLibraryPublicClass.Run1", StringComparison.Ordinal))
        {
            var instanceType = types["ClassLibrary8.FooLibraryPublicClass"];
            var i = _functionActivator.CreateInstance(instanceType, context) as global::ClassLibrary8.FooLibraryPublicClass;
            context.GetInvocationResult().Value = i.Run1((global::Microsoft.Azure.Functions.Worker.Http.HttpRequestData)inputArguments[0]);
        }
    }

    private IFunctionExecutor CreateDefaultExecutorInstance(FunctionContext context)
    {
        var defaultExecutorFullName = "Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor, Microsoft.Azure.Functions.Worker.Core, Version=1.16.0.0, Culture=neutral, PublicKeyToken=551316b6919f366c";
        var defaultExecutorType = Type.GetType(defaultExecutorFullName);

        return ActivatorUtilities.CreateInstance(context.InstanceServices, defaultExecutorType) as IFunctionExecutor;
    }
}
```

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [x] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: sdk/Sdk.Generators/FunctionMethodVisibility.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Microsoft.Azure.Functions.Worker.Sdk.Generators
+{
+    /// <summary>
+    /// Represents the visibility of a "azure function" method and it's parent classes.
+    /// </summary>
+    internal enum FunctionMethodVisibility
+    {
+        /// <summary>
+        /// The method and it's parent classes are public.
```
Comment (satvu) [inline sdk/Sdk.Generators/FunctionMethodVisibility.cs:12]:
```suggestion
        /// The method and its parent classes are public.
```

File: sdk/Sdk.Generators/FunctionMethodVisibility.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.Functions.Worker.Sdk.Generators
+{
+    /// <summary>
+    /// Represents the visibility of a "azure function" method and it's parent classes.
```
Comment (satvu) [inline sdk/Sdk.Generators/FunctionMethodVisibility.cs:7]:
```suggestion
    /// Represents the visibility of an "Azure function" method and it's parent classes.
```

File: sdk/Sdk.Generators/FunctionMethodVisibility.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    /// </summary>
+    internal enum FunctionMethodVisibility
+    {
+        /// <summary>
+        /// The method and it's parent classes are public.
+        /// </summary>
+        PublicAndVisible,
+
+        /// <summary>
+        /// The method is public, but one or more of it's parent classes are not public.
```
Comment (satvu) [inline sdk/Sdk.Generators/FunctionMethodVisibility.cs:17]:
```suggestion
        /// The method is public, but one or more of its parent classes are not public.
```

File: sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.Emitter.cs [hunk 4]
```diff
@@ -67,20 +69,40 @@ public static IHostBuilder ConfigureGeneratedFunctionExecutor(this IHostBuilder
...
+                    """;
+            }
+
             private static string GetTypesDictionary(IEnumerable<ExecutableFunction> functions)
             {
-                var classNames = functions.Where(f => !f.IsStatic).Select(f => f.ParentFunctionClassName).Distinct();
-                if (!classNames.Any())
-                {
-                    return """
+                // Build a dictionary of type names and it's full qualified names (including assembly identity)
```
Comment (satvu) [inline sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.Emitter.cs:91]:
```suggestion
                // Build a dictionary of type names and its full qualified names (including assembly identity)
```

File: sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.Emitter.cs [hunk 5]
```diff
@@ -67,20 +69,40 @@ public static IHostBuilder ConfigureGeneratedFunctionExecutor(this IHostBuilder
                 return result;
             }
 
+            private static string EmitCreateDefaultExecutorMethod(GeneratorExecutionContext context)
+            {
+                var workerCoreAssembly = context.Compilation.SourceModule.ReferencedAssemblySymbols.Single(a => a.Name == "Microsoft.Azure.Functions.Worker.Core");
```
Comment (satvu) [inline sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.Emitter.cs:74]: Can some of the strings in this class be constants (in this file or another)?

File: sdk/Sdk.Generators/FunctionMethodVisibility.cs (hunk 2 above)
Comment (kshyju) [reply sdk/Sdk.Generators/FunctionMethodVisibility.cs]: Fixed.

File: sdk/Sdk.Generators/FunctionMethodVisibility.cs (hunk 3 above)
Comment (kshyju) [reply sdk/Sdk.Generators/FunctionMethodVisibility.cs]: Fixed.

File: sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.Emitter.cs (hunk 4 above)
Comment (kshyju) [reply sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.Emitter.cs]: Fixed.

File: sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.Emitter.cs (hunk 5 above)
Comment (kshyju) [reply sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.Emitter.cs]: moved assembly name to a constant.

File: sdk/Sdk.Generators/MethodVisibilityChecker.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.Azure.Functions.Worker.Sdk.Generators
+{
+    internal static class MethodVisibilityChecker
```
Comment (liliankasem) [inline sdk/Sdk.Generators/MethodVisibilityChecker.cs:8]: Could this be defined as an extension class  i.e. `MethodSymbolExtensions` where  GetVisibility looks like this: `GetVisibility(this IMethodSymbol methodSymbol)`

File: sdk/Sdk.Generators/FunctionMethodVisibility.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        /// <summary>
+        /// The method and it's parent classes are public.
+        /// </summary>
+        PublicAndVisible,
+
+        /// <summary>
+        /// The method is public, but one or more of its parent classes are not public.
+        /// </summary>
+        PublicButContainingTypeNotVisible,
```
Comment (liliankasem) [inline sdk/Sdk.Generators/FunctionMethodVisibility.cs:19]: nit:  `Public`, `NotPublic` and `PublicWithNonVisibleTypes` 

Or does it make more sense to have a more common enum i,.e.

`Public`, `Private`, `Internal` etc.

File: sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.cs [hunk 8]
```diff
@@ -23,25 +26,52 @@ public void Execute(GeneratorExecutionContext context)
...
+        /// <summary>
+        /// Collect methods with Function attributes on them from dependent/referenced assemblies.
+        /// </summary>
+        private static IEnumerable<IMethodSymbol> GetDependentAssemblyFunctionsSymbols(GeneratorExecutionContext context)
+        {
+            var visitor = new ReferencedAssemblyMethodVisitor(context.Compilation);
+            visitor.Visit(context.Compilation.SourceModule);
+
+            return visitor.FunctionMethods;
+        }
```
Comment (liliankasem) [inline sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.cs:74]: nit: new line between methods

```suggestion
        }
 
```

File: sdk/Sdk.Generators/FunctionMethodVisibility.cs (hunk 7 above)
Comment (liliankasem) [inline sdk/Sdk.Generators/FunctionMethodVisibility.cs:19]: Curious why this is treated differently / needs to be its own state

File: sdk/Sdk.Generators/MethodVisibilityChecker.cs (hunk 6 above)
Comment (kshyju) [reply sdk/Sdk.Generators/MethodVisibilityChecker.cs]: Moved to an extension method.

File: sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.cs (hunk 8 above)
Comment (kshyju) [reply sdk/Sdk.Generators/FunctionExecutor/FunctionExecutorGenerator.cs:74]: Fixed line break.

File: sdk/Sdk.Generators/FunctionMethodVisibility.cs (hunk 7 above)
Comment (kshyju) [reply sdk/Sdk.Generators/FunctionMethodVisibility.cs:19]: If one writes a public method, but the class it resides is internal, we want to treat them differently when executing. We will not be able to use the src gen version executor approach in that case as we will not be able to access the type from the entry assembly. In that case, we will fallback to the existing reflection based executor. 
The legacy version of executor supports this use case. So our new executor  (src gen version) should also work for that case.

Comment (kshyju) [reply sdk/Sdk.Generators/FunctionMethodVisibility.cs:19]: Changed PublicAndVisible to Public.  I still feel "PublicButContainingTypeNotVisible" is more readable. So sticking with that.

File: sdk/release_notes.md [hunk 9]
```diff
@@ -4,12 +4,14 @@
 - My change description (#PR/#issue)
 -->
 
-### Microsoft.Azure.Functions.Worker.Sdk 1.16.4 (meta package)
-
+### Microsoft.Azure.Functions.Worker.Sdk 1.16.3 (meta package)
```
Comment (kshyju) [inline sdk/release_notes.md:7]: We haven't released 1.16.3 yet.

**Review by liliankasem — APPROVED** [review]

**Review by satvu — APPROVED** [review]

**Pull request merged**